_posts/2016-08-10-dynamic-loading.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Dynamic Code Loading on a MCU
+authors:
+  - brghena
+  - alevy
+---
+
+One key feature of Tock is the ability to load and run multiple applications
+simultaneously. In a modern computer, the OS uses the Memory Management Unit
+(MMU) to provide a virtual address space for each process. The code for each
+process is written assuming that it is the only code in the world and that it
+can place memory at any address it pleases. The MMU handles translating these
+virtual addresses into a physical address in real memory, which is shared
+between all processes. Unfortunately, in the world of embedded systems MMUs are
+not available. Processors like the ARM Cortex-M series omit them because they
+aren't necessary for traditional single-process embedded devices.
```

Comment:
I think there's actually an interesting argument to be made that MMUs do not inherently require such large pages, rather that's a necessary granularity given the total size of memory in modern systems to keep the page table metadata size in check, both in total memory and hw units to support it (c.f. modern hugs pages for further evidence of this). I'm not convinced that page size is actually a compelling argument against MMUs for MCUs, but that argument is orthogonal to the current discussion.